### ISSUE TITLE
Automatically wrap list of target TF ops with nvtx plugin

### ISSUE THREAD

**Issue by shanesyy-1992** [NVIDIA/nvtx-plugins#21]
Hi, thanks a lot for the great job! I wonder is there a way to automatically wrap all the target TF ops with nvtx start and end, like modifying the create_op function in tf.Graph, rather than using nvtx_tf.ops.start and nvtx_tf.ops.end to wrap each one by one? Thanks!

**Comment by ahmadki** [comment]
If you are looking to tag TF **operations**, NGC containers wraps them for you automatically and you don't need to use nvtx-plugins. TF containers: https://ngc.nvidia.com/catalog/containers/nvidia:tensorflow

If you are looking to tag **layers**, you can monkey patch the TF base layer. For example, if you are using keras API:

```python
from tensorflow.python.keras.engine.base_layer import Layer
from nvtx.plugins.tf.keras.layers import NVTXStart, NVTXEnd

old_call = Layer.__call__

def new_layer_call(self, inputs, **kwargs):
    inputs, marker_id, domain_id = NVTXStart(message=self.name,
                                             domain_name='forward')(inputs)
    outputs = old_call(self, inputs, **kwargs)
    outputs = NVTXEnd(grad_message=self.name + ' grad',
                      grad_domain_name='backwards')([outputs, marker_id, domain_id])
    return outputs

Layer.__call__ = new_layer_call
```
Note that the first layer in your model will need special handling where `NVTXStart` needs to be called with `trainable=True`

**Comment by shanesyy-1992** [comment]
Hi Ahmad,
Thanks a lot for your detailed response. The layer wrapper does work! 
Just a quick question, what if I set all `NVTXStart` with `trainable=True`? Would that greatly affect the training performance? Thanks!

**Comment by ahmadki** [comment]
`NVTXStart` and `NVTXEnd` don't have internal weights and therefore not trainable by default. This is an issue for us because TF will optimize your graph and choose not to propagate gradients through the layers if it doesn't need to (like when `NVTXStart` is the first layer), resulting in open-ended traces in the gradient domain.

`trainable=True` simply adds a dummy weight to the layer to trick TF into thinking the layer is trainable, the dummy weight is not used by `NVTXStart` so its impact is minimal. However, you are still adding an extra weight that TF needs to "optimize".

It's impossible to estimate the performance impact without knowing the model you are using, but in most cases it should be negligible. If you do notice a large performance hit with `trainable=True`, let us know and we'll try to find a better solution.

**Comment by shanesyy-1992** [comment]
Understand now. Thank you so much for your response, learned a lot! I'll report the performance issue with`trainable=True` if it happens. Thanks!

**Comment by ahmadki** [comment]
@shanesyy-1992 please close the issue if you have no further questions.
Thank you
